hdr/sqlite_modern_cpp.h

```diff
@@ -16,515 +17,480 @@
+		}
 
-class database;
-class database_binder;
+		static void throw_custom_error(const char* str) {
+			throw std::runtime_error(str);
+		}
 
-template<std::size_t> class binder;
+	}
 
-template<typename T> database_binder&& operator <<(database_binder&& db,T const&& val);
-template<typename T> void get_col_from_db(database_binder& db, int inx, T& val);
+	class database;
+	class database_binder;
 
-class sqlite3_statement
-{
-	private:
-		sqlite3_stmt* _stmt;
+	template<std::size_t> class binder;
 
-	public:
-		
-		sqlite3_stmt** operator&()
-		{
-			return &_stmt;
-		}
+	typedef std::shared_ptr<sqlite3> connection_type;
 
-		operator sqlite3_stmt*()
-		{
-			return _stmt;
+	template<typename Tuple, int Element = 0, bool Last = (std::tuple_size<Tuple>::value == Element)> struct tuple_iterate {
+		static void iterate(Tuple& t, database_binder& db) {
+			get_col_from_db(db, Element, std::get<Element>(t));
+			tuple_iterate<Tuple, Element + 1>::iterate(t, db);
 		}
+	};
 
-		sqlite3_statement(sqlite3_stmt* s)
-		:_stmt(s)
-		{
-		}
+	template<typename Tuple, int Element> struct tuple_iterate<Tuple, Element, true> {
+		static void iterate(Tuple&, database_binder&) {}
+	};
+
+	class database_binder {
+
+	public:
+		typedef std::unique_ptr<sqlite::database_binder> chain_type;
 
-		~sqlite3_statement()
-		{
-			//Do not check for errors: an error code means that the 
-			//*execution* of the statement failed somehow. We deal with errors
-			//at that point so we don't need to know about errors here.
-			//
-			//Also, this is an RAII class to make sure we don't leak during exceptions
-			//so there's a reasonably chance we're already in an exception here.
-			
-			sqlite3_finalize(_stmt);
+		void reset() {
+			sqlite3_reset(_stmt.get());
+			sqlite3_clear_bindings(_stmt.get());
+			_inx = 1;
 		}
-};
 
-template<typename Tuple, int Element=0, bool Last=(std::tuple_size<Tuple>::value == Element)> struct tuple_iterate
-{
-	static void iterate(Tuple& t, database_binder& db)
-	{
-		get_col_from_db(db,Element, std::get<Element>(t));
-		tuple_iterate<Tuple, Element+1>::iterate(t, db);
-	}
-};
+		void execute() {
+			int hresult;
 
-template<typename Tuple, int Element> struct tuple_iterate<Tuple, Element, true>
-{
-	static void iterate(Tuple&, database_binder&)
-	{
-	}
-};
-
-class database_binder {
-private:
-	sqlite3* const _db;
-	std::u16string _sql;
-	sqlite3_statement _stmt;
-	int _inx;
-	
-	bool execution_started = false;
-	bool throw_exceptions = true;
-	bool error_occured = false;
-
-	void _extract(std::function<void(void)> call_back) {
-		execution_started = true;
-		int hresult;
+			while((hresult = sqlite3_step(_stmt.get())) == SQLITE_ROW) {}
 
-		while ((hresult = sqlite3_step(_stmt)) == SQLITE_ROW) {
-			call_back();
+			if(hresult != SQLITE_DONE) {
+				exceptions::throw_sqlite_error(hresult);
+			}
 		}
 
-		if (hresult != SQLITE_DONE) {
-			throw_sqlite_error(hresult);
-		}
+		void set_used(bool state) { execution_started = state; }
```

Comment:
instead of two functions overload `.used() // getter` and `.used(bool) // setter` functions